Update NPathComplexity Check to Account for when expression as a Possible Execution Path

check doc: https://checkstyle.org/checks/metrics/npathcomplexity.html#Description

---

```
D:\CS\test javac src/Test.java
D:\CS\test cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="NPathComplexity">
      <property name="max" value="0"/>
    </module>
  </module>
</module>
D:\CS\test cat src/Test.java
public class Test {

    // violation, NPath Complexity is 3
    void m(Object o) {
        if (o instanceof String s && !s.isEmpty()) {

        } else {

        }
    }

    // violation, NPath Complexity is 2
    void m2(Object o) {
        switch (o) {
            case String s when !s.isEmpty() -> { }
            default -> {}
        }
    }
}
D:\CS\test java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:4:5: NPath Complexity is 3 (max allowed is 0). [NPathComplexity]
[ERROR] D:\CS\test\src\Test.java:13:5: NPath Complexity is 2 (max allowed is 0). [NPathComplexity]
Audit done.
Checkstyle ends with 2 errors.
```
---

The current implementation of the NPath Complexity Check does not account for the complexity added by `when`. We should update the NPath Complexity Check to add a complexity point for `when`.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java
@@ -91,6 +91,7 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * <td>NP(for-range) + NP(expr1)+ NP(expr2) + NP(expr3) + 1</td></tr>
  * <tr><td>switch ([expr]) { case : [case-range] default: [default-range] }</td>
  * <td>S(i=1:i=n)NP(case-range[i]) + NP(default-range) + NP(expr)</td></tr>
+ * <tr><td>when[expr]</td><td>NP(expr) + 1</td></tr>
  * <tr><td>[expr1] ? [expr2] : [expr3]</td><td>NP(expr1) + NP(expr2) + NP(expr3) + 2</td></tr>
  * <tr><td>goto label</td><td>1</td></tr><tr><td>break</td><td>1</td></tr>
  * <tr><td>Expressions</td>
@@ -220,6 +221,7 @@ public final class NPathComplexityCheck extends AbstractCheck {
             TokenTypes.LITERAL_DEFAULT,
             TokenTypes.COMPACT_CTOR_DEF,
             TokenTypes.SWITCH_RULE,
+            TokenTypes.LITERAL_WHEN,
         };
     }
 
@@ -248,6 +250,9 @@ public final class NPathComplexityCheck extends AbstractCheck {
                 break;
             case TokenTypes.LITERAL_RETURN:
                 visitUnitaryOperator(ast, 0);
+                break;
+            case TokenTypes.LITERAL_WHEN:
+                visitWhenExpression(ast, 1);
                 break;
             case TokenTypes.CASE_GROUP:
                 final int caseNumber = countCaseTokens(ast);
@@ -291,6 +296,7 @@ public final class NPathComplexityCheck extends AbstractCheck {
             case TokenTypes.LITERAL_FOR:
             case TokenTypes.LITERAL_IF:
             case TokenTypes.LITERAL_SWITCH:
+            case TokenTypes.LITERAL_WHEN:
                 leaveConditional();
                 break;
             case TokenTypes.LITERAL_TRY:
@@ -340,6 +346,19 @@ public final class NPathComplexityCheck extends AbstractCheck {
             expressionValue += countConditionalOperators(bracketed);
         }
         processingTokenEnd.setToken(bracketed);
+        pushValue(expressionValue);
+    }
+
+    /**
+     * Visits when expression token. There is no guarantee that when expression will be
+     * bracketed, so we don't use visitConditional method.
+     *
+     * @param ast visited token.
+     * @param basicBranchingFactor default number of branches added.
+     */
+    private void visitWhenExpression(DetailAST ast, int basicBranchingFactor) {
+        final int expressionValue = basicBranchingFactor + countConditionalOperators(ast);
+        processingTokenEnd.setToken(getLastToken(ast));
         pushValue(expressionValue);
     }
 

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/metrics/NPathComplexityCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/metrics/NPathComplexityCheck.xml
@@ -64,6 +64,7 @@
  &lt;td&gt;NP(for-range) + NP(expr1)+ NP(expr2) + NP(expr3) + 1&lt;/td&gt;&lt;/tr&gt;
  &lt;tr&gt;&lt;td&gt;switch ([expr]) { case : [case-range] default: [default-range] }&lt;/td&gt;
  &lt;td&gt;S(i=1:i=n)NP(case-range[i]) + NP(default-range) + NP(expr)&lt;/td&gt;&lt;/tr&gt;
+ &lt;tr&gt;&lt;td&gt;when[expr]&lt;/td&gt;&lt;td&gt;NP(expr) + 1&lt;/td&gt;&lt;/tr&gt;
  &lt;tr&gt;&lt;td&gt;[expr1] ? [expr2] : [expr3]&lt;/td&gt;&lt;td&gt;NP(expr1) + NP(expr2) + NP(expr3) + 2&lt;/td&gt;&lt;/tr&gt;
  &lt;tr&gt;&lt;td&gt;goto label&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;td&gt;break&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;/tr&gt;
  &lt;tr&gt;&lt;td&gt;Expressions&lt;/td&gt;

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheckTest.java
@@ -257,8 +257,8 @@ public class NPathComplexityCheckTest extends AbstractModuleTestSupport {
             "23:5: " + getCheckMessage(MSG_KEY, 3, 1),
             "32:5: " + getCheckMessage(MSG_KEY, 3, 1),
             "41:5: " + getCheckMessage(MSG_KEY, 3, 1),
-            "50:5: " + getCheckMessage(MSG_KEY, 3, 1),
-            "59:5: " + getCheckMessage(MSG_KEY, 3, 1),
+            "50:5: " + getCheckMessage(MSG_KEY, 5, 1),
+            "59:5: " + getCheckMessage(MSG_KEY, 5, 1),
             "68:5: " + getCheckMessage(MSG_KEY, 4, 1),
             "76:5: " + getCheckMessage(MSG_KEY, 4, 1),
             "86:5: " + getCheckMessage(MSG_KEY, 3, 1),
@@ -267,6 +267,28 @@ public class NPathComplexityCheckTest extends AbstractModuleTestSupport {
 
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputNPathComplexityPatternMatchingForSwitch.java"),
+            expected);
+
+    }
+
+    @Test
+    public void testWhenExpression() throws Exception {
+
+        final String[] expected = {
+            "14:5: " + getCheckMessage(MSG_KEY, 3, 1),
+            "20:5: " + getCheckMessage(MSG_KEY, 3, 1),
+            "28:5: " + getCheckMessage(MSG_KEY, 3, 1),
+            "36:5: " + getCheckMessage(MSG_KEY, 4, 1),
+            "44:5: " + getCheckMessage(MSG_KEY, 4, 1),
+            "52:5: " + getCheckMessage(MSG_KEY, 5, 1),
+            "60:5: " + getCheckMessage(MSG_KEY, 7, 1),
+            "69:5: " + getCheckMessage(MSG_KEY, 5, 1),
+            "77:5: " + getCheckMessage(MSG_KEY, 5, 1),
+            "85:5: " + getCheckMessage(MSG_KEY, 6, 1),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputNPathComplexityWhenExpression.java"),
             expected);
 
     }
@@ -294,6 +316,7 @@ public class NPathComplexityCheckTest extends AbstractModuleTestSupport {
             TokenTypes.LITERAL_DEFAULT,
             TokenTypes.COMPACT_CTOR_DEF,
             TokenTypes.SWITCH_RULE,
+            TokenTypes.LITERAL_WHEN,
         };
         assertWithMessage("Acceptable tokens should not be null")
             .that(actual)
@@ -326,6 +349,7 @@ public class NPathComplexityCheckTest extends AbstractModuleTestSupport {
             TokenTypes.LITERAL_DEFAULT,
             TokenTypes.COMPACT_CTOR_DEF,
             TokenTypes.SWITCH_RULE,
+            TokenTypes.LITERAL_WHEN,
         };
         assertWithMessage("Required tokens should not be null")
             .that(actual)

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/npathcomplexity/InputNPathComplexityPatternMatchingForSwitch.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/npathcomplexity/InputNPathComplexityPatternMatchingForSwitch.java
@@ -46,7 +46,7 @@ public class InputNPathComplexityPatternMatchingForSwitch {
         }
     }
 
-    // violation below, 'NPath Complexity is 3 (max allowed is 1)'
+    // violation below, 'NPath Complexity is 5 (max allowed is 1)'
     void testGuardsInRule(Object o) {
         switch (o) {
             case Integer i when i > 0 -> {}
@@ -55,7 +55,7 @@ public class InputNPathComplexityPatternMatchingForSwitch {
         }
     }
 
-    // violation below, 'NPath Complexity is 3 (max allowed is 1)'
+    // violation below, 'NPath Complexity is 5 (max allowed is 1)'
     void testGuardsInStatement(Object o) {
         switch (o) {
             case Integer i when i > 0 : {} break;

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/npathcomplexity/InputNPathComplexityWhenExpression.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/npathcomplexity/InputNPathComplexityWhenExpression.java
@@ -1,0 +1,91 @@
+/*
+NPathComplexity
+max = 1
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.metrics.npathcomplexity;
+
+public class InputNPathComplexityWhenExpression {
+
+    // violation below, 'NPath Complexity is 3 (max allowed is 1)'
+    void m(Object o) {
+
+        if (o instanceof String s && !s.isEmpty()) { }
+    }
+
+    // violation below, 'NPath Complexity is 3 (max allowed is 1)'
+    void m2(Object o) {
+        switch (o) {
+            case String s when !s.isEmpty() -> { }
+            default -> { }
+        }
+    }
+
+    // violation below, 'NPath Complexity is 3 (max allowed is 1)'
+    void m3(Object o) {
+        switch (o) {
+            case String s when (!s.isEmpty()) -> { }
+            default -> { }
+        }
+    }
+
+    // violation below, 'NPath Complexity is 4 (max allowed is 1)'
+    void m4(Object o, boolean b) {
+        switch (o) {
+            case String s when !s.isEmpty() && b -> { }
+            default -> { }
+        }
+    }
+
+    // violation below, 'NPath Complexity is 4 (max allowed is 1)'
+    void m5(Object o, boolean b) {
+        switch (o) {
+            case String s when (!s.isEmpty() && b) -> { }
+            default -> { }
+        }
+    }
+
+    // violation below, 'NPath Complexity is 5 (max allowed is 1)'
+    void m6(Object o, boolean b, boolean c) {
+        switch (o) {
+            case String s when (!s.isEmpty() && b) && c -> { }
+            default -> { }
+        }
+    }
+
+    // violation below, 'NPath Complexity is 7 (max allowed is 1)'
+    void m7(Object o, boolean b, boolean c) {
+        switch (o) {
+            case String s when !s.isEmpty() -> { }
+            case Integer i when (i == 0) && c || b -> { }
+            default -> { }
+        }
+    }
+
+    // violation below, 'NPath Complexity is 5 (max allowed is 1)'
+    void m8(Object o, boolean b, boolean c) {
+        switch (o) {
+            case String s when (b ? true : c) -> { }
+            default -> { }
+        }
+    }
+
+    // violation below, 'NPath Complexity is 5 (max allowed is 1)'
+    void m9(Object o, boolean b, boolean c) {
+        switch (o) {
+            case String s when b ? true : c -> { }
+            default -> { }
+        }
+    }
+
+    // violation below, 'NPath Complexity is 6 (max allowed is 1)'
+    void m10(Object o, boolean b, boolean c) {
+        switch (o) {
+            case String s when (b ? true : c) && c == true -> { }
+            default -> { }
+        }
+    }
+}

--- a/src/xdocs/checks/metrics/npathcomplexity.xml
+++ b/src/xdocs/checks/metrics/npathcomplexity.xml
@@ -75,6 +75,7 @@
             <tr><td>switch ([expr]) { case : [case-range] default: [default-range] }</td>
               <td>S(i=1:i=n)NP(case-range[i]) + NP(default-range) + NP(expr)</td>
             </tr>
+            <tr><td>when[expr]</td><td>NP(expr) + 1</td></tr>
             <tr><td>[expr1] ? [expr2] : [expr3]</td><td>NP(expr1) + NP(expr2) + NP(expr3) + 2
             </td></tr>
             <tr><td>goto label</td><td>1</td></tr>

--- a/src/xdocs/checks/metrics/npathcomplexity.xml.template
+++ b/src/xdocs/checks/metrics/npathcomplexity.xml.template
@@ -75,6 +75,7 @@
             <tr><td>switch ([expr]) { case : [case-range] default: [default-range] }</td>
               <td>S(i=1:i=n)NP(case-range[i]) + NP(default-range) + NP(expr)</td>
             </tr>
+            <tr><td>when[expr]</td><td>NP(expr) + 1</td></tr>
             <tr><td>[expr1] ? [expr2] : [expr3]</td><td>NP(expr1) + NP(expr2) + NP(expr3) + 2
             </td></tr>
             <tr><td>goto label</td><td>1</td></tr>

